core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -194,189 +216,149 @@ class TransactionCoordinator(brokerId: Int,
                                        pid: Long,
                                        epoch: Short,
                                        partitions: collection.Set[TopicPartition],
-                                       responseCallback: TxnMetadataUpdateCallback): Unit = {
-    val errors = validateTransactionalId(transactionalId)
-    if (errors != Errors.NONE)
-      responseCallback(errors)
-    else {
+                                       responseCallback: AddPartitionsCallback): Unit = {
+    val error = validateTransactionalId(transactionalId)
+    if (error != Errors.NONE) {
+      responseCallback(error)
+    } else {
       // try to update the transaction metadata and append the updated metadata to txn log;
       // if there is no such metadata treat it as invalid pid mapping error.
-      val (error, newMetadata) = txnManager.getTransactionState(transactionalId) match {
+      val result: Either[Errors, (Int, TransactionMetadataTransition)] = txnManager.getTransactionState(transactionalId) match {
         case None =>
-          (Errors.INVALID_PID_MAPPING, null)
+          Left(Errors.INVALID_PID_MAPPING)
+
+        case Some(epochAndMetadata) =>
+          val coordinatorEpoch = epochAndMetadata.coordinatorEpoch
+          val txnMetadata = epochAndMetadata.transactionMetadata
 
-        case Some(metadata) =>
           // generate the new transaction metadata with added partitions
-          metadata synchronized {
-            if (metadata.pid != pid) {
-              (Errors.INVALID_PID_MAPPING, null)
-            } else if (metadata.producerEpoch != epoch) {
-              (Errors.INVALID_PRODUCER_EPOCH, null)
-            } else if (metadata.pendingState.isDefined) {
+          txnMetadata synchronized {
+            if (txnMetadata.producerId != pid) {
+              Left(Errors.INVALID_PID_MAPPING)
+            } else if (txnMetadata.producerEpoch != epoch) {
+              Left(Errors.INVALID_PRODUCER_EPOCH)
+            } else if (txnMetadata.pendingTransitionInProgress) {
               // return a retriable exception to let the client backoff and retry
-              (Errors.CONCURRENT_TRANSACTIONS, null)
-            } else if (metadata.state == PrepareCommit || metadata.state == PrepareAbort) {
-              (Errors.CONCURRENT_TRANSACTIONS, null)
+              Left(Errors.CONCURRENT_TRANSACTIONS)
+            } else if (txnMetadata.state == PrepareCommit || txnMetadata.state == PrepareAbort) {
+              Left(Errors.CONCURRENT_TRANSACTIONS)
```

Comment:
Well, since we return to the user right after we have appended the log, in which case the state has updated to `PrepareXX`, the client code is possible to send another `addPartitions` while the sending markers are still on the flight. In this case we should not return a fatal error but let client retry.